sopel/plugins/rules.py

```diff
@@ -1301,3 +1358,140 @@ def parse(self, text):
             match = regex.search(text)
             if match:
                 yield match
+
+
+class URLCallback(Rule):
+    """URL callback rule definition.
+
+    A URL callback rule (or simply "a URL rule") detects URLs in a trigger
+    then it uses regular expressions to match at most once per URL per regular
+    expression, i.e. you can trigger between 0 and the number of regex the URL
+    callback has per URL in the IRC line.
+
+    Here is an example with a URL rule with the pattern
+    ``r'https://example\\.com/(.*)'``:
+
+    .. code-block:: irc
+
+        <user> https://example.com/test
+        <Bot> You triggered a URL callback, with the "/test" path
+        <user> and this URL is https://example.com/other can you get it?
+        <Bot> You triggered a URL callback, with the "/other" path
+
+    Like generic rules, URL callback rules are not triggered by any specific
+    name and they don't have aliases.
+
+    .. note::
+
+        Unlike generic rules and commands, the :func:`~sopel.plugin.url`
+        decorator expects its decorated function to have the bot and the
+        trigger with a third parameter: the ``match`` parameter.
+
+        To use this class with an existing URL callback handler, the
+        :meth:`from_callable` classmethod **must** be used: it will wrap the
+        handler to work as intended. In that case, the ``trigger`` and the
+        ``match`` arguments will be the same when the rule executes.
+
+        This behavior makes the ``match`` parameter obsolete.
```

Comment:
If we elect to mark `match` as deprecated and plan to remove it in 9.0 (or whenever), a note should be added here too.